lib/services/api/api_service.dart

```diff
@@ -0,0 +1,89 @@
+import 'package:survey/gen/configs.gen.dart';
+import 'package:survey/services/http/http_service.dart';
+import 'package:object_mapper/object_mapper.dart';
+import 'package:survey/services/locator/locator_service.dart';
+
+part 'api_exception.dart';
+
+part 'api_params.dart';
+
+part 'api_service_register.dart';
+
+abstract class ApiService {
+  Future<T> call<T extends Mappable>({
+    required HttpMethod method,
+    String? baseUrl,
+    required String endPoint,
+    ApiParams? params,
+    String? token,
+  });
+
+  void configureGlobalBaseUrl(String? baseUrl);
+
+  void configureGlobalToken(String? token, String? tokenType);
+}
+
+class ApiServiceImpl implements ApiService {
+  static String? _baseUrl;
+  static String? _token;
+  static String _tokenType = "Bearer";
+  final HttpService _httpService = locator.get();
+
+  @override
+  Future<T> call<T extends Mappable>({
+    required HttpMethod method,
+    String? baseUrl,
+    required String endPoint,
+    ApiParams? params,
+    String? token,
+    String? tokenType,
+  }) async {
+    final String? finalBaseUrl = baseUrl ?? _baseUrl;
+    assert(finalBaseUrl != null);
+
+    final Map<String, dynamic> headers = {};
+
+    // Append bearer token
+    final String? finalToken = token ?? _token;
+    if (finalToken != null) {
+      final finalTokenType = tokenType ?? _tokenType;
+      headers["authorization"] = "$finalTokenType $finalToken";
+    }
+
+    final url = finalBaseUrl! + endPoint;
+    try {
+      final response = await _httpService.request(
+        method: method,
+        data: params?.toJson(),
+        url: url,
+        headers: headers,
+      ) as Map<String, dynamic>;
+      return Mapper.fromJson(
+              response["data"]["attributes"] as Map<String, dynamic>)
+          .toObject<T>();
+    } on HttpException catch (e) {
+      throw ApiException.fromHttpException(e) ?? e;
+    }
+  }
+
+  /// Configures a default base url when [call] with `baseUrl` is null.
+  ///
+  /// This value is stored in a static variable [_baseUrl] so it will be available in any
+  /// instance of [ApiServiceImpl], after set
+  @override
+  void configureGlobalBaseUrl(String? baseUrl) {
+    _baseUrl = baseUrl;
+  }
+
+  /// Configures a default token & tokenType when [call] with token or tokenType is null.
+  ///
+  /// This value is stored in a static variable [_token], [_tokenType] so it will be available in any
+  /// instance of [ApiServiceImpl], after set
+  @override
+  void configureGlobalToken(String? token, String? tokenType) {
+    _token = token;
+    if (tokenType != null) {
+      _tokenType = tokenType;
+    }
+  }
```

Comment:
@markgravity I think this is not the best approach for calling every endpoints with saved token, imagine the scenario that we already logged in (have token) then we call an endpoint that doesn't need `authorization`, but we still add `authorization` for that endpoint, it is redundant imo 🤔  Or we have to set `token` and `tokenType` to be `null` before calling the endpoint, this is still not good. Please correct me if I'm wrong 🙏

Reply:
@manh-t 
You're right. Let's me add a flag to ignore them like this
```dart
  Future<T> call<T extends Mappable>({
     required HttpMethod method,
     String? baseUrl,
     required String endPoint,
     ApiParams? params,
     bool requiresAuthentication = true, // <- new flag
     String? token,
     String? tokenType,
   })
```

Reply:
@markgravity yes, we can do that for current implementation 🙆‍♂️  That's nice 🙏

Reply:
I often create 2 instances to handle auth and non-auth, this is the other approach too 😁